core/src/scene.cpp

```diff
@@ -115,5 +124,40 @@ std::optional<GameObject> Scene::get_game_object_from_component(T& component) {
     entt::entity handle = entt::to_entity(scene.m_registry, component);
     return scene.get_game_object_from_handle(handle);
 }
+void Scene::save_scene_to_stream(std::ostream& serialized) {
+    cereal::JSONOutputArchive archive(serialized);
+    // TODO: When you make a new component add it here to the snapshot if it needs to be
+    entt::snapshot{m_registry}
+        .entities(archive)
+        .component<uuid, components::Name, components::Tag, components::Transform, components::Hierarchy,
+                   components::Material, components::InstanceMesh, components::SpotLight, components::EditorCamera>(
+            archive);
+}
+void Scene::load_scene_from_stream(std::istream& serialized) {
+    m_uuidGameObjectMap.clear();
+    m_entityGameObjectMap.clear();
+
+    cereal::JSONInputArchive archive(serialized);
+    // TODO: When you make a new component add it here to the snapshot if it needs to be
+    entt::snapshot_loader{m_registry}
+        .entities(archive)
+        .component<uuid, components::Name, components::Tag, components::Transform, components::Hierarchy,
+                   components::Material, components::InstanceMesh, components::SpotLight, components::EditorCamera>(
+            archive);
+    auto view = m_registry.view<uuid>();
+    for (auto ent : view) {
+        add_game_object(ent);
```

Comment:
Why don't we first add the game objects, then load the rest of the components?
Genuine question.

Reply:
the snapshot loader has already loaded the entities and all their components. This creates the gameObjects that sit on top of each entity and adds them to the maps in the scene, so they can be accessed in other parts of the engine

Reply:
Definition:
  - Guaranteed Components: Components added when `create_game_object`, Hierarchy, Name, and Transform.
  
If a component's `on_awake` function needs a `Transform` component, for example, or 
depends on any other component, is it guaranteed that you would have already loaded the guaranteed components? Or what is the order of component loading?

Reply:
Because the loader has to have an identical list to the snapshot, including ordering you can order components within the list dependent on this - hence we have all guaranteed components first, so if they're needed in on_awakes later this can happen. If we find this isn't working properly we can call the on_awakes for all components needed to after the snapshot loading is complete. At the moment the split serialize with on_awake is only use by assets. I'm assuming it will need to be used in physics but hopefully this should just work?

If not then we'll try it this other way

Reply:
Okay, perfect!